server/src/main/java/org/opensearch/action/support/TransportAction.java

```diff
@@ -155,18 +156,25 @@ public void onFailure(Exception e) {
      * Use this method when the transport action should continue to run in the context of the current task
      */
     public final void execute(Task task, Request request, ActionListener<Response> listener) {
-        ActionRequestValidationException validationException = request.validate();
-        if (validationException != null) {
-            listener.onFailure(validationException);
-            return;
-        }
+        ThreadContext.StoredContext storedContext = taskManager.addTaskIdInThreadContext(task);
```

Comment:
We should add task in thread context centrally as a part of task creation, instead of doing it at multiple places after task creation, it could lead to maintenance overhead overtime

Reply:
This is done as per the ThreadContext's usage documentation which is something like :
1. Clone/Stash existing thread context
2. Add new fields to the new thread context. 
3. Execute the operation (or take task to a different thread)
4. Restore the original thread context. 


If we want to do it in register method then we would have to face challenges of making ThreadContext mutable which today throws exception and will also need to find a common place to remove the header from the thread context so that some other request doesn't use this just because we missed to remove it from thread context.

Reply:
Lets just call out this works for tasks that forks off on a separate thread(non-network thread)

Reply:
@tushar-kharbanda72 I would like to back @Bukhtawar concerns, manipulating thread context in a such away is not desirable (cause essentially someone down the line could override it again). The suggestion I have in mind is to introduce the distinction between `Runnable` (no task context) and `TaskRunnable extends Runnable` (with Task ID as final property). In this case, the taskId could be extracted by thread pool and baked into the thread local context from the start (by checking if `Runnable` is instance of `TaskRunnable`). What do you think?

Reply:
@tushar-kharbanda72 yes, it is possible to manage task id transfer using stash(), but fundamental problem here (in my opinion) - the thread context should not be manipulated through `TaskManager` from the outside. Yes, it will introduce larger scope of changes, the benefit - less dependency on thread context (black box), explicit link between task and its execution model, just my 5 cents, @Bukhtawar do you have an option?

Reply:
> fundamental problem here (in my opinion) - the thread context should not be manipulated through TaskManager from the outside

I agree. This is essentially using the thread context to stash a global variable and is likely to lead to pain in the long term.

Reply:
@reta @andrross Currently coupling the thread context update with when a task gets registered in resource tracking service as both together made more sense. 

Having task id set as a member in the runnable is an ideal case scenario but comes with it's own sets of challenges:
- Submitting a runnable will not be very flexible. We would need Task Id available as a function param at every point throughout the code base wherever we need to submit a runnable. 
- If a blocking sync child action gets executed on the same thread - this would essentially mean updating the runnable's task id in some way and reverting it back as soon as it finishes. 
- For new features folks would need to handle the overlapping case of making task id available and I see this would increase the number of params they handle in the function. 

I see the benefits of the suggested approach but it also comes with the above mentioned downsides and these could be even more when we deep dive further. 

Also, similar approach is used for request tracing `X-Opaque-Id` in OpenSearch which I see a more critical feature for some users.


@Bukhtawar Would like to hear your thoughts on this as well.

Reply:
@tushar-kharbanda72 sorry for the delay, I certainly agree it is challenging to have robust and reliable context propagation support but this is also very important to avoid any surprises by opening the gate for (arbitrary) context manipulations. The approach could be generalized by eliminating the focus on task to abstract runnable with contextual setup: in this case the thread context could be established by runnable (fe, associated with task or anything else) before hand. 

There are quite a few examples of contextual runnables like `ContextPreservingAbstractRunnable`  and `ContextPreservingRunnable`. 

`X-Opaque-Id` is also a good example of context propagation but it is passed through `requestHeaders`, may be we could extend `ThreadContextStruct` with non-headers key/value pairs, task id being one of them. I will try to prototype the possible implementation(s) shortly and share with you.

Reply:
@reta Created an issue to discuss further on this https://github.com/opensearch-project/OpenSearch/issues/2819#issue-1197334468. We can use this to close the final change required for making task id available to runnable.

I believe this can be added in a follow up PR as it's not changing the core logic and just the way runnable knows about the Task. Do let me know your thoughts.